java/runtime/src/main/java/org/ray/runtime/serializer/MessagePackSerializer.java

```diff
@@ -0,0 +1,245 @@
+        if (type.isAssignableFrom(byte[].class)) {
+          return v.asBinaryValue().asByteArray();
+        } else {
+          throw new IllegalArgumentException("expected " + type + ", actual byte[]!");
+        }
+      case ARRAY:
+        if (type.isArray() || type.isAssignableFrom(Object.class)) {
+          ArrayValue a = v.asArrayValue();
+          Class<?> componentType = type.isArray() ? type.getComponentType() : Object.class;
+          Object array = Array.newInstance(componentType, a.size());
+          for (int i = 0; i < a.size(); ++i) {
+            Value value = a.get(i);
+            Array.set(array, i, unpack(value, componentType, javaDeserializer));
+          }
+          return array;
+        } else {
+          throw new IllegalArgumentException("expected " + type + ", actual Array!");
+        }
+      case EXTENSION:
+        ExtensionValue ev = v.asExtensionValue();
+        byte extType = ev.getType();
+        if (extType == CROSS_LANGUAGE_TYPE_EXTENSION_ID) {
+          try {
+            MessageUnpacker unpacker = MessagePack.newDefaultUnpacker(ev.getData());
+            Value crossValue = unpacker.unpackValue();
+            // data is [type id, cross data object]
+            Object[] data = (Object[]) unpack(crossValue, Object[].class, javaDeserializer);
+            Preconditions.checkState(data != null && data.length == 2);
+            Integer crossTypeId = ((Number) data[0]).intValue();
+            Object[] crossData = (Object[]) data[1];
+            // Get type by type id and call KEY_FROM_CROSS_DATA method with cross data.
+            Class<?> crossType = CrossTypeManager.get(crossTypeId);
+            Method fromCrossData = crossType.getDeclaredMethod(
+                CrossTypeManager.KEY_FROM_CROSS_DATA, Object[].class);
+            fromCrossData.setAccessible(true);
+            return fromCrossData.invoke(null, new Object[]{crossData});
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        } else if (extType == LANGUAGE_SPECIFIC_TYPE_EXTENSION_ID) {
+          return javaDeserializer.deserialize(ev);
+        } else {
+          throw new IllegalArgumentException("expected " + type + ", actual extension type " +
+              extType);
+        }
+      default:
+        throw new IllegalArgumentException(
+            "expected " + type + ", actual type " + v.getValueType());
+    }
+  }
+
+  public static byte[] encode(Object obj, Serializer.Meta meta, ClassLoader classLoader) {
+    MessageBufferPacker packer = MessagePack.newDefaultBufferPacker();
+    try {
+      // Reserve MESSAGE_PACK_OFFSET bytes for MessagePack bytes length.
+      packer.writePayload(new byte[MESSAGE_PACK_OFFSET]);
+      // Serialize input object by MessagePack.
+      Serializer.Meta javaEncoderMeta = new Serializer.Meta();
+      pack(obj, packer, ((object, packer1) -> {
+        byte[] payload = FSTSerializer.encode(object, javaEncoderMeta, classLoader);
```

Comment:
2 small issues:
1) It'd be nicer if `MessagePackSerializer` is decoupled with `FSTSerializer`. 
2) `Serializer.Meta` is used as an output argument, which is not a common usage in Java and the name is a bit ambiguous. 

I think maybe we can change the code in this way:
1) Let `Serializer` pass in `javaSerializer` function into `MessagePackSerializer`.
2) `javaSerializer` just calls `FSTSerializer.encode`, without doing anything related to message pack.
3) `javaSerializer` should also hold a boolean wrapper, so we know if it's called. (similar to `Meta` but maybe use a different name).
4) `Serializer` returns a pair of byte[] and boolean.

Reply:
1, I think it's better to make the output of Serializer can be the input of Deserializer.
2, If we add  `javaSerializer` to MessagePackSerializer, we can't switch serializer between  MessagePackSerializer and FSTSerializer smoothly. Currently, the interfaces of MessagePackSerializer and FSTSerializer are identical.

Reply:
Your points make sense as well. My main concern was that these 3 serializers are too tightly coupled. It's not clear what is the boundary of each one.

Reply:
Another way is to merge `Serializer.java` with `ObjectSerializer.java` and use `NativeRayObject` as Serializer output?
Then, we can use `MessagePackSerializer` as the main serializer, `FSTSerializer` as one choice of `javaSerializer`.